### DISCUSSION TITLE
Flask JWT Usage

### DISCUSSION THREAD

**Discussion by tsungchiehchen** [ResilientApp/Res-a-Me#53] [General]
### Frontend:

Every component can get `access_token` using sessionStorage.getItem('access_token').
`access_token` can be included in the Header with `Bearer Token`  `Authorization` using `GET` method to send the request to the backend server to verify whether the user is currently logged in or not.

The following example can be located in /frontend/src/router/router.js

``` javascript
const response = await fetch("http://127.0.0.1:3033/loadUser", {
       method: "GET",
       headers: {
               "Content-type": "application/json; charset=UTF-8",
               "Access-Control-Allow-Origin": "*",
               "Authorization": `Bearer ` + sessionStorage.getItem('access_token'),
       }
});
```
`Response code: 200` => user is authenticated
`Response code: 401` => user is not authenticated

For any component who need to check whether a user has already logged in or not, can make the same request as above but place that in your component's mounted() method.

The following example can be located in /frontend/src/vue/core/LandingPage.vue

``` javascript
<script>
export default {
  // add this mounted() method if you don't have
  mounted() {
    fetch("http://127.0.0.1:3033/loadUser", {
        method: "GET",
        headers: {
          "Content-type": "application/json; charset=UTF-8",
          "Authorization": `Bearer ` + sessionStorage.getItem('access_token'),
        }
    })
    .then((response) => response.json())
    .then((json) => {
        if (json.status === 200) {
            console.log("User has logged in")
            console.log("Logged in user's email: ", json.logged_in_as)
        }
        else {
            this.errorMessage = "User are not logged in";
        }
    })
    .catch(error => {
        console.error('There was an error!', error);
        this.errorMessage = error.message || "An error occurred. Please try again.";
    });
  },
  methods:{
    signin(){
      this.$router.push('/login')
    }
  },
  // .... remaining lines are the same
};
</script>
```

### Backend:

Place @jwt_required() before each protected API.
Within the API function calls, use get_jwt_identity() to get the current user email. Then, you have successfully retrieve user's identity.

The following example can be located in /backend/main.py

``` python
# Protect a route with jwt_required
# kick out requests without a valid JWT present
@app.route("/loadUser", methods=["GET"])
@jwt_required()
def load_logged_in_user():
    # Access the identity of the current user with get_jwt_identity
    current_user = get_jwt_identity()
    if current_user:
        return jsonify(logged_in_as=current_user, status = 200)
    else:
        return jsonify(status = 401)
```

### Additional Information

![rt-and-at](https://github.com/tobyyu007/Res-a-Me/assets/21220686/7b82f007-1e16-436d-a81b-8b551c259148)

SPA = Single-Page Application; AS = Authorization Server; RS = Resource Server; AT = Access Token; RT = Refresh Token.

In our case, our AS is the same as RS.

If there is no RT, then once the AT has expired, user will need to re-login. Therefore, once the AT has expired, the SPA will make a request with the RT to AS to get a new AT. Then, SPA will make a new request with new AT to the RS.

**Comment by tsungchiehchen** [comment]
BTW, the code can be found in Login_Fullstack branch

**Comment by tobyyu007** [comment]
圖文並茂 XDD
